packages/wrangler/src/cloudchamber/deploy.ts

```diff
@@ -0,0 +1,168 @@
+import { readFileSync, statSync } from "fs";
+import path from "path";
+import { type Config } from "../config";
+import { type ContainerApp } from "../config/environment";
+import { UserError } from "../errors";
+import { CI } from "../is-ci";
+import { type Logger } from "../logger";
+import { fetchVersion } from "../versions/api";
+import { apply } from "./apply";
+import { build } from "./build";
+import { fillOpenAPIConfiguration } from "./common";
+
+export async function buildContainers(config: Config, workerTag: string) {
+	if (config.containers === undefined) {
+		return;
+	}
+
+	for (const container of config.containers) {
+		const options = getBuildArguments(container, workerTag, true);
+		if (options.isDockerImage) {
+			await build(options);
+		}
+	}
+}
+
+export async function deployContainers(
+	logger: Logger,
+	config: Config,
+	{
+		versionId,
+		accountId,
+		scriptName,
+		dryRun,
+		env,
+	}: {
+		versionId: string;
+		accountId: string;
+		scriptName: string;
+		dryRun?: boolean;
+		env?: string;
+	}
+) {
+	if (config.containers === undefined) {
+		return;
+	}
+
+	if (!dryRun) {
+		await fillOpenAPIConfiguration(config, CI.isCI());
+	}
+
+	for (const container of config.containers) {
+		const version = await fetchVersion(accountId, scriptName, versionId);
+		const targetDurableObject = version.resources.bindings.filter(
+			(durableObject) =>
+				durableObject.type === "durable_object_namespace" &&
+				durableObject.class_name === container.class_name &&
+				durableObject.script_name === undefined &&
+				durableObject.namespace_id !== undefined
+		);
+
+		if (targetDurableObject.length <= 0) {
+			logger.error(
+				"Could not deploy container application as durable object was not found in list of bindings"
+			);
+			continue;
+		}
+
+		const [targetDurableObjectNamespace] = targetDurableObject;
+		if (
+			targetDurableObjectNamespace.type !== "durable_object_namespace" ||
+			targetDurableObjectNamespace.namespace_id === undefined
+		) {
+			throw new Error("unreachable");
+		}
+
+		const configuration = {
+			...config,
+			containers: [
+				{
+					...container,
+					durable_objects: {
+						namespace_id: targetDurableObjectNamespace.namespace_id,
+					},
+				},
+			],
+		};
+
+		const buildOptions = getBuildArguments(container, versionId, dryRun);
+
+		if (buildOptions.isDockerImage) {
+			logger.log("Building image", buildOptions.tag);
+		}
+
+		const image = buildOptions.isDockerImage
+			? await build(buildOptions)
+			: container.image ?? container.configuration.image;
+
+		container.configuration.image = image;
+		container.image = image;
+
+		await apply({ skipDefaults: false, json: true, env }, configuration);
+	}
+}
+
+function getBuildArguments(
+	container: ContainerApp,
+	versionId: string,
+	dryRun?: boolean
+) {
+	let isDockerImage = false;
+	const imageRef = container.image ?? container.configuration.image;
+
+	let imagePath = path.resolve(imageRef);
```

Comment:
Could this potentially fail/break if `imageRef` is a URL at this point?

Reply:
No, it won't fail if that's the case. We have tested being able to pass a URI image.